src/wallet/wallet.cpp

```diff
@@ -2880,21 +2880,31 @@ CBitcoinAddress CWallet::ParseIntoAddress(const CTxDestination& dest, const std:
     return CBitcoinAddress(dest, addrType);
 }
 
+/**
+ * TODO: If we ever add new functionality that modifies the purpose of an address book entry
+ * outside of the below situations, we need to also ensure those situations are considered here
+ */
+bool CWallet::UpdatablePurpose(const std::string& strPurpose1, const std::string& strPurpose2)
+{
+    return (strPurpose1 == AddressBook::AddressBookPurpose::DELEGABLE && strPurpose2 == AddressBook::AddressBookPurpose::DELEGATOR) ||
+           (strPurpose1 == AddressBook::AddressBookPurpose::DELEGATOR && strPurpose2 == AddressBook::AddressBookPurpose::DELEGABLE);
+}
+
 bool CWallet::SetAddressBook(const CTxDestination& address, const std::string& strName, const std::string& strPurpose)
 {
     bool fUpdated = HasAddressBook(address);
+    const bool fUpdatePurpose = !fUpdated || UpdatablePurpose(mapAddressBook.at(address).purpose, strPurpose);
     {
         LOCK(cs_wallet); // mapAddressBook
         mapAddressBook[address].name = strName;
```

Comment:
Can use `mapAddressBook.at` instead of `[]` for consistency here.

Reply:
This was left as-is because in the event that the address is not already in the address book, the `[]` operator will also insert a new element to the map automatically. Up until this point we only access the map for read purposes, and only if we already know that the address exists in the map.